sql/stmtdiagnostics: TestDiagnosticsRequest failed

sql/stmtdiagnostics.TestDiagnosticsRequest [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10802099?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10802099?buildTab=artifacts#/) on release-23.1 @ [4c7b13ec11aac7910182d645f03080c5c3c1b80e](https://github.com/cockroachdb/cockroach/commits/4c7b13ec11aac7910182d645f03080c5c3c1b80e):


```
=== RUN   TestDiagnosticsRequest
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/cb4e70d2cc4a901d2a7689654ffd0100/logTestDiagnosticsRequest164094824
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestDiagnosticsRequest
    statement_diagnostics_test.go:62: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/sql/stmtdiagnostics_test/pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:62
        	            				github.com/cockroachdb/cockroach/pkg/sql/stmtdiagnostics_test/pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:82
        	            				github.com/cockroachdb/cockroach/pkg/sql/stmtdiagnostics_test/pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:327
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:59
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41
        	            				github.com/cockroachdb/cockroach/pkg/sql/stmtdiagnostics_test/pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:296
        	Error:      	Should be false
        	Test:       	TestDiagnosticsRequest
        	Messages:   	request was completed and should have been removed from the registry
=== CONT  TestDiagnosticsRequest
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/cb4e70d2cc4a901d2a7689654ffd0100/logTestDiagnosticsRequest164094824
--- FAIL: TestDiagnosticsRequest (10.86s)
=== RUN   TestDiagnosticsRequest/ongoing_request_canceled
=== CONT  TestDiagnosticsRequest/ongoing_request_canceled
    testing.go:1343: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
    --- FAIL: TestDiagnosticsRequest/ongoing_request_canceled (0.53s)
=== RUN   TestDiagnosticsRequest/ongoing_request_canceled/conditional
=== CONT  TestDiagnosticsRequest/ongoing_request_canceled/conditional
    testing.go:1343: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
        --- FAIL: TestDiagnosticsRequest/ongoing_request_canceled/conditional (0.53s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDiagnosticsRequest.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29482

--- a/pkg/sql/instrumentation.go
+++ b/pkg/sql/instrumentation.go
@@ -408,6 +408,9 @@ func (ih *instrumentationHelper) Finish(
 		)
 		phaseTimes := statsCollector.PhaseTimes()
 		execLatency := phaseTimes.GetServiceLatencyNoOverhead()
+		// Note that we want to remove the request from the local registry
+		// _before_ inserting the bundle to prevent rare test flakes (#106284).
+		ih.stmtDiagnosticsRecorder.MaybeRemoveRequest(ih.diagRequestID, ih.diagRequest, execLatency)
 		if ih.stmtDiagnosticsRecorder.IsConditionSatisfied(ih.diagRequest, execLatency) {
 			placeholders := p.extendedEvalCtx.Placeholders
 			ob := ih.emitExplainAnalyzePlanToOutputBuilder(ih.explainFlags, phaseTimes, queryLevelStats)
@@ -430,7 +433,6 @@ func (ih *instrumentationHelper) Finish(
 			)
 			telemetry.Inc(sqltelemetry.StatementDiagnosticsCollectedCounter)
 		}
-		ih.stmtDiagnosticsRecorder.MaybeRemoveRequest(ih.diagRequestID, ih.diagRequest, execLatency)
 	}
 
 	// If there was a communication error already, no point in setting any

